srml/scored-pool/src/lib.rs

```diff
@@ -0,0 +1,370 @@
+		pub fn kick(origin, who: T::AccountId) {
+			T::KickOrigin::try_origin(origin)
+				.map(|_| ())
+				.or_else(ensure_root)
+				.map_err(|_| "bad origin")?;
+
+			Self::remove_member(who)?;
+			Self::deposit_event(RawEvent::CandidateKicked);
+		}
+
+		/// Score a member `who` with `score`.
+		///
+		/// May only be called from `ScoreOrigin` or root.
+		pub fn score(origin, who: T::AccountId, score: T::Score) {
+			T::ScoreOrigin::try_origin(origin)
+				.map(|_| ())
+				.or_else(ensure_root)
+				.map_err(|_| "bad origin")?;
+
+			let mut pool = <Pool<T, I>>::get();
+			let location = Self::find_in_pool(&who)?;
+			pool.remove(location);
+
+			// we binary search the pool (which is sorted by score).
+			// if there is already an element with `score`, we insert
+			// right before that. if not, the search returns a location
+			// where we can insert while maintaining order.
+			let item = (who.clone(), Some(score.clone()));
+			let location = pool
+				.binary_search_by_key(
+					&Some(score),
+					|(_, score)| score.clone(),
+				)
+				.unwrap_or_else(|l| l);
+			pool.insert(location, item);
+
+			<Pool<T, I>>::put(&pool);
+		}
+
+		/// Root-dispatchable call to change `MemberCount`.
+		pub fn change_member_count(origin, count: u32) {
+			ensure_root(origin)?;
+			<MemberCount<I>>::put(&count);
+		}
+	}
+}
+
+impl<T: Trait<I>, I: Instance> Module<T, I> {
+
+	/// Fetches the `MemberCount` highest scoring members from
+	/// `Pool` and puts them into `Members`.
+	///
+	/// If `notify` is set to true `T::MembershipChanged::change_members_sorted`
+	/// will be invoked at the end of the method.
+	fn refresh_members(notify: bool) {
+		let count = <MemberCount<I>>::get();
+
+		let pool = <Pool<T, I>>::get();
+		let mut new_members: Vec<T::AccountId> = pool
+			.into_iter()
+			.filter(|(_, score)| score.is_some())
+			.rev()
+			.take(count as usize)
+			.map(|(account_id, _)| account_id)
+			.collect();
+		new_members.sort();
+
+		let old_members = <Members<T, I>>::get();
+		<Members<T, I>>::put(&new_members);
+
+		let outgoing: Vec<T::AccountId> = old_members.clone()
+			.into_iter()
+			.filter(|old| new_members.binary_search(&old).is_err())
+			.collect();
+		let incoming: Vec<T::AccountId> = new_members.clone()
+			.into_iter()
+			.filter(|new| old_members.binary_search(&new).is_err())
+			.collect();
+
+		if notify {
+			T::MembershipChanged::change_members_sorted(
+				&incoming[..],
+				&outgoing[..],
+				&new_members[..]
+			);
+		}
+	}
+
+	/// Sorts the `Pool` by score in an ascending order. Entities which
+	/// have a score of `None` are sorted to the beginning of the vec.
+	fn sort_pool(pool: &mut Vec<(T::AccountId, Option<T::Score>)>) {
+		pool.sort_by_key(|(_, maybe_score)| maybe_score.unwrap_or_default());
+	}
+
+	/// Find an entity in the pool.
+	///
+	/// Returns its position in the `Pool` vec, if found.
+	fn find_in_pool(seek: &T::AccountId) -> Result<usize, &'static str> {
```

Comment:
rename to `ensure_in_pool` to make its meaning clear

Reply:
There is a `ensure!(Self::find_in_pool(&who).is_err(), "already a member");` which takes use of this function, but besides that there is another use of the function which actually needs the `usize` location in the vec. So I think the name still fits and would leave it as is.

Reply:
in that case avoid returning a string error.

Reply:
actually, just remove this function altogether. it's way too dangerous.